pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java

```diff
@@ -0,0 +1,198 @@
+      _tableNamesWithoutType = val;
+      return this;
+    }
+
+    /**
+     * set the threshold for _numEntriesScannedInFilter, the queries with _numEntriesScannedInFilter below this will be filtered out
+     * @param val
+     * @return
+     */
+    @Nonnull
+    public Builder setNumEntriesScannedThreshold(long val) {
+      _numEntriesScannedThreshold = val;
+      return this;
+    }
+
+    /**
+     * set the cardinality threshold, column with cardinality below this will be ignored,
+     * setting a high value will force the system to ignore low card columns
+     * @param val cardinality threshold, default to 1
+     * @return
+     */
+    @Nonnull
+    public Builder setCardinalityThreshold(long val) {
+      _cardinalityThreshold = val;
+      return this;
+    }
+
+    /**
+     * set the minimum number of records scanned to give a recommendation
+     * @param val minimum number of records scanned to give a recommendation, default to 0
+     * @return
+     */
+    @Nonnull
+    public Builder setNumQueriesThreshold(long val) {
+      _numQueriesThreshold = val;
+      return this;
+    }
+  }
+
+  @Override
+  public boolean filter(AbstractQueryStats queryStats) {
+    IndexSuggestQueryStatsImpl indexSuggestQueryStatsImpl = (IndexSuggestQueryStatsImpl) queryStats;
+    long numEntriesScannedInFilter = Long.parseLong(indexSuggestQueryStatsImpl.getNumEntriesScannedInFilter());
+    return (_tableNamesWithoutType == null || _tableNamesWithoutType.isEmpty() || _tableNamesWithoutType
+        .contains(indexSuggestQueryStatsImpl.getTableNameWithoutType())) && (numEntriesScannedInFilter
+        > _numEntriesScannedThreshold);
+  }
+
+  @Override
+  public void accumulate(AbstractQueryStats queryStats, MetaManager metaManager,
+      Map<String, Map<String, AbstractAccumulator>> accumulatorOut) {
+
+    IndexSuggestQueryStatsImpl indexSuggestQueryStatsImpl = (IndexSuggestQueryStatsImpl) queryStats;
+    String tableNameWithoutType = indexSuggestQueryStatsImpl.getTableNameWithoutType();
+    String numEntriesScannedInFilter = indexSuggestQueryStatsImpl.getNumEntriesScannedInFilter();
+    String query = indexSuggestQueryStatsImpl.getQuery();
+    LOGGER.debug("Accumulator: scoring query {}", query);
+    HashSet<String> counted = new HashSet<>();
+
+    accumulatorOut.putIfAbsent(tableNameWithoutType, new HashMap<>());
+    accumulatorOut.get(tableNameWithoutType).putIfAbsent(NUM_QUERIES_COUNT, new ParseBasedAccumulator());
+    accumulatorOut.get(tableNameWithoutType).get(NUM_QUERIES_COUNT).increaseCount();
+
+    Matcher matcher = _dimensionPattern.matcher(query);
+    while (matcher.find()) {
+      if (matcher.group(1) != null) {
+        counted.add(matcher.group(1));
+      } else if (matcher.group(4) != null) {
+        counted.add(matcher.group(4));
+      } else {
+      }
+    }
+
+    counted.stream().filter(colName -> metaManager.getColumnSelectivity(tableNameWithoutType, colName)
```

Comment:
even the simple frequency based requires the metadata manager? no option to skip it?

Reply:
This is slightly more intelligent than very simple frequency, we can set a cardinality threshold for the simple counter, which is more reasonable